### PR TITLE
web/admin: easy ALG/STUN guesser

### PR DESCRIPTION
**Pull request by cruzccl**
#### Type Of Change 
- [ ] Small bug fix
- [x] New feature or enhancement
- [ ] Breaking change

#### Checklist:
- [x] Commits are named and have tag following [commit rules](https://github.com/irontec/ivozprovider/blob/bleeding/doc/dev/en/commits.md)
- [x] Commits are split per component (schema, web/admin, kamusers, agis, ..)
- [x] Changes have been tested locally
- [ ] Fixes an existing issue (Fixes #XXXX) 

#### Description

Add a hint line in register status that tries to guess awkward registrations:

- Public IP address in REGISTER Contact (SIP ALG/STUN)
- Private address in both Contact and Received (internal routing)

#### Additional information

Nothing serious, just a humble guess.

### CODE REVIEW

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 1]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+    {
+        $ctc = $this->isRFC1918($contact) ? 'private' : 'public';
+
+        if (is_null($received)) {
+            if ($ctc === 'private') {
+                $comment = "No NAT with private Contact (hint: internal routing)";
+            } else {
+                $comment = "No NAT with public Contact (hint: SIP ALG / STUN)";
+            }
+        } else {
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:383]: with returns there is not need of else

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 2]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+     * @param $received
+     * @return string
+     */
+    private function getRegisterHint($contact, $received) : string
+    {
+        $ctc = $this->isRFC1918($contact) ? 'private' : 'public';
+
+        if (is_null($received)) {
+            if ($ctc === 'private') {
+                $comment = "No NAT with private Contact (hint: internal routing)";
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:379]: return 'No NAT with p...'

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 3]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+            } else {
+                $comment = "No NAT with public Contact (hint: SIP ALG / STUN)";
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:381]: return "No NAT with pu....

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 4]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+            }
+        } else {
+            $rcv = $this->isRFC1918($received) ? 'private' : 'public';
+
+            if ($ctc === 'private' && $rcv === 'public') {
+                $comment = "Regular NAT detected (private Contact, public Received)";
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:387]: return...

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 5]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+            } else {
+                $comment = "Awkward NAT detected ($ctc Contact, $rcv Received)";
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:389]: return

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 6]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+     * Check if given source in IP(:PORT) format is private
+     *
+     * @param $src
+     * @return bool
+     */
+    private function isRFC1918($src) : bool
+    {
+        list ($ip, $port) = explode(':', $src); // Extract address if port is given
+
+        $pri_addrs = array (
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:406]: camel case

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 7]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+            '10.0.0.0|10.255.255.255', // single class A network
+            '172.16.0.0|172.31.255.255', // 16 contiguous class B network
+            '192.168.0.0|192.168.255.255', // 256 contiguous class C network
+            '169.254.0.0|169.254.255.255', // Link-local address also refered to as Automatic Private IP Addressing
+            '127.0.0.0|127.255.255.255' // localhost
+        );
+
+        $long_ip = ip2long($ip);
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:414]: camel case

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 8]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+        if ($long_ip != -1) {
+            foreach ($pri_addrs as $pri_addr) {
+                list ($start, $end) = explode('|', $pri_addr);
+
+                if ($long_ip >= ip2long($start) && $long_ip <= ip2long($end)) {
```
Comment (Kaian) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:419]: camelCase variables here

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 9]
```diff
@@ -360,4 +364,64 @@ private function getLocationStatus($model)
...
+            }
+        } else {
+            $rcv = $this->isRFC1918($received) ? 'private' : 'public';
+
+            if ($ctc === 'private' && $rcv === 'public') {
+                $comment = "Regular NAT detected (private Contact, public Received)";
+            } else {
+                $comment = "Awkward NAT detected ($ctc Contact, $rcv Received)";
+            }
+        }
```
Comment (Kaian) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:391]: what about this code 

```suggestion
        $contactIsPrivate = $this->isRFC1918($contact);

        if (is_null($received)) {
            if ($contactIsPrivate) {
                $comment = "No NAT with private Contact (hint: internal routing)";
            } else {
                $comment = "No NAT with public Contact (hint: SIP ALG / STUN)";
            }
        } else {
            $receivedIsPublic = $this->isRFC1918($received);

            if ($contactIsPrivate && $receivedIsPublic) {
                $comment = "Regular NAT detected (private Contact, public Received)";
            } else {
                $comment = "Awkward NAT detected (" 
                          . $contactIsPrivate ?  "Private" : "Public"
                          . "Contact, "
                          . $receivedIsPublic ?  "Public" : "Private"
                          . "Received)";
            }
        }
```

Use variables with more letters forming words in PHP



```

```

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 10]
```diff
@@ -360,4 +364,68 @@ private function getLocationStatus($model)
...
+     */
+    private function getRegisterHint($contact, $received) : string
+    {
+        $contactIsPrivate = $this->isRFC1918($contact);
+
+        // No NAT
+        if (is_null($received)) {
+            if ($contactIsPrivate) {
+                return "No NAT with private Contact (hint: internal routing)";
+            } else {
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:381]: no need of else

File: web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php [hunk 11]
```diff
@@ -360,4 +364,68 @@ private function getLocationStatus($model)
...
+                return "No NAT with public Contact (hint: SIP ALG / STUN)";
+            }
+        }
+
+        // Behind NAT
+        $receivedIsPublic = !$this->isRFC1918($received);
+
+        if ($contactIsPrivate && $receivedIsPublic) {
+            return "Regular NAT detected (private Contact, public Received)";
+        } else {
```
Comment (mmadariaga) [inline web/admin/application/library/IvozProvider/Klear/Ghost/RegisterStatus.php:391]: no need of else

**Pull request merged**